packages/firestore/src/local/index_backfiller.ts

```diff
@@ -0,0 +1,216 @@
+ */
+import { DocumentMap } from '../model/collections';
+import {
+  IndexOffset,
+  indexOffsetComparator,
+  newIndexOffsetFromDocument
+} from '../model/field_index';
+import { debugAssert } from '../util/assert';
+import { AsyncQueue, DelayedOperation, TimerId } from '../util/async_queue';
+import { logDebug } from '../util/log';
+
+import { INDEXING_ENABLED } from './indexeddb_schema';
+import { ignoreIfPrimaryLeaseLoss, LocalStore } from './local_store';
+import { LocalWriteResult } from './local_store_impl';
+import { Persistence, Scheduler } from './persistence';
+import { PersistencePromise } from './persistence_promise';
+import { PersistenceTransaction } from './persistence_transaction';
+import { isIndexedDbTransactionError } from './simple_db';
+
+const LOG_TAG = 'IndexBackiller';
+
+/** How long we wait to try running index backfill after SDK initialization. */
+const INITIAL_BACKFILL_DELAY_MS = 15;
+
+/** Minimum amount of time between backfill checks, after the first one. */
+const REGULAR_BACKFILL_DELAY_MS = 1;
+
+/** The maximum number of documents to process each time backfill() is called. */
+const MAX_DOCUMENTS_TO_PROCESS = 50;
+
+/** This class is responsible for the scheduling of Index Backfiller. */
+export class IndexBackfillerScheduler implements Scheduler {
+  private task: DelayedOperation<void> | null;
+
+  constructor(
+    private readonly asyncQueue: AsyncQueue,
+    private readonly backfiller: IndexBackfiller
+  ) {
+    this.task = null;
+  }
+
+  start(): void {
+    debugAssert(
+      this.task === null,
+      'Cannot start an already started IndexBackfillerScheduler'
+    );
+    if (INDEXING_ENABLED) {
+      this.schedule(INITIAL_BACKFILL_DELAY_MS);
+    }
+  }
+
+  stop(): void {
+    if (this.task) {
+      this.task.cancel();
+      this.task = null;
+    }
+  }
+
+  get started(): boolean {
+    return this.task !== null;
+  }
+
+  private schedule(delay: number): void {
+    debugAssert(
+      this.task === null,
+      'Cannot schedule IndexBackiller while a task is pending'
+    );
+    logDebug(LOG_TAG, `Scheduled in ${delay}ms`);
+    this.task = this.asyncQueue.enqueueAfterDelay(
+      TimerId.IndexBackfill,
+      delay,
+      async () => {
+        this.task = null;
+        try {
+          const documentsProcessed = await this.backfiller.backfill();
+          logDebug(LOG_TAG, `Documents written: ${documentsProcessed}`);
+        } catch (e) {
+          if (isIndexedDbTransactionError(e)) {
+            logDebug(
+              LOG_TAG,
+              'Ignoring IndexedDB error during index backfill: ',
+              e
+            );
+          } else {
+            await ignoreIfPrimaryLeaseLoss(e);
+          }
+        }
+        await this.schedule(REGULAR_BACKFILL_DELAY_MS);
+      }
+    );
+  }
+}
+
+/** Implements the steps for backfilling indexes. */
+export class IndexBackfiller {
+  constructor(
+    private readonly localStore: LocalStore,
```

Comment:
Add a comment to explain why we need localstore here.

Reply:
Comment added